### PR TITLE
Allow ref/pull format in install_dev

### PR DESCRIPTION
**Pull request by mkearney**
As requested in #206. Checks and then stores ref/pull format to include in install_repo call.

### CODE REVIEW

**Comment by gaborcsardi** [comment]
Thanks! I wonder if we could do this with a single regex. E.g. sg like 
```
re_match(c("foo", "bar", "foobar@ref"), "^(?<pkg>[^@#]+)(?<ref>[@#].*)?$")
```
will split up the input, and then you just add back the ref, like you already do. I would put this in a `dev_split_ref()` function, so that we can test it.

**Comment by mkearney** [comment]
Sounds good! I'm on it!

File: R/install-dev.R [hunk 1]
```diff
@@ -28,15 +28,7 @@ install_dev <- function(package, cran_url = getOption("repos")[["CRAN"]], ...) {
...
-  # if @[ref] is included, save for repo install
-  if (contains_ref(package)) {
-    ref <- sub(".*(?=(@|#))", "", package, perl = TRUE)
-    package <- sub("(#|@).*", "", package)
-  } else {
-    ref <- ""
-  }
-
-  url <- build_url(cran_url, "web", "packages", package, "DESCRIPTION")
+  url <- build_url(cran_url, "web", "packages", dev_split_ref(package)[["pkg"]], "DESCRIPTION")
```
Comment (mkearney) [inline R/install-dev.R]: I don't know if it's the best way to access "pkg" and "ref" values, but I used double brackets and column names.

File: tests/testthat/test-utils.R [hunk 2]
```diff
@@ -179,3 +179,15 @@ test_that("in_r_build_ignore works", {
...
+  expect_equal(dev_split_ref("DT")[["pkg"]], "DT")
+  expect_equal(dev_split_ref("remotes")[["ref"]], "")
+  expect_equal(dev_split_ref("with.dot")[["pkg"]], "with.dot")
+  expect_equal(dev_split_ref("with2")[["pkg"]], "with2")
+  expect_equal(dev_split_ref("with@v1.2.1")[["ref"]], "@v1.2.1")
+  expect_equal(dev_split_ref("with@v1.0.0.999")[["ref"]], "@v1.0.0.999")
+  expect_equal(dev_split_ref("with@v1.0.0.999")[["pkg"]], "with")
+  expect_equal(dev_split_ref("with#279")[["ref"]], "#279")
+  expect_equal(dev_split_ref("with#1")[["pkg"]], "with")
+})
```
Comment (mkearney) [inline tests/testthat/test-utils.R]: These tests all assume input format is `(pkg[(@ref|#pull)]?)`

File: R/install-dev.R (hunk 1 above)
Comment (gaborcsardi) [reply R/install-dev.R]: Yes, that's good, but can we call `dev_split_ref` only once? Not a big a deal performance-wise, more of a principle.  Store the result here, and use it later below.

File: tests/testthat/test-utils.R (hunk 2 above)
Comment (gaborcsardi) [reply tests/testthat/test-utils.R]: Yes, these  look great!

**Comment by jimhester** [comment]
Great work, this package is now better thanks to you!

Thanks!! You are da bomb!

**Pull request merged**